script/tool/lib/src/format_command.dart

```diff
@@ -79,66 +91,105 @@ class FormatCommand extends PluginCommand {
-    await processRunner.runAndStream('java',
-        <String>['-jar', googleFormatterPath, '--replace', ...javaFiles],
-        workingDir: packagesDir, exitOnError: true);
+  Future<void> _formatJava(
+      Iterable<String> files, String googleFormatterPath) async {
+    final Iterable<String> javaFiles =
+        _getPathsWithExtensions(files, <String>{'.java'});
+    if (javaFiles.isNotEmpty) {
+      print('Formatting .java files...');
+      final int exitCode = await processRunner.runAndStream('java',
+          <String>['-jar', googleFormatterPath, '--replace', ...javaFiles],
+          workingDir: packagesDir);
+      if (exitCode != 0) {
+        printError('Failed to format Java files: exit code $exitCode.');
+        throw ToolExit(_exitJavaFormatFailed);
+      }
+    }
   }
 
-  Future<void> _formatDart() async {
-    // This actually should be fine for non-Flutter Dart projects, no need to
-    // specifically shell out to dartfmt -w in that case.
-    print('Formatting all .dart files...');
-    final Iterable<String> dartFiles = await _getFilesWithExtension('.dart');
-    if (dartFiles.isEmpty) {
-      print(
-          'No .dart files to format. If you set the `--exclude` flag, most likey they were skipped');
-    } else {
-      await processRunner.runAndStream(
+  Future<void> _formatDart(Iterable<String> files) async {
+    final Iterable<String> dartFiles =
+        _getPathsWithExtensions(files, <String>{'.dart'});
+    if (dartFiles.isNotEmpty) {
+      print('Formatting .dart files...');
+      // `flutter format` doesn't require the project to actually be a Flutter
+      // project.
+      final int exitCode = await processRunner.runAndStream(
           'flutter', <String>['format', ...dartFiles],
-          workingDir: packagesDir, exitOnError: true);
+          workingDir: packagesDir);
+      if (exitCode != 0) {
+        printError('Failed to format Dart files: exit code $exitCode.');
+        throw ToolExit(_exitFlutterFormatFailed);
+      }
     }
   }
 
-  Future<List<String>> _getFilesWithExtension(String extension) async =>
-      getFiles()
-          .where((File file) => p.extension(file.path) == extension)
-          .map((File file) => file.path)
-          .toList();
+  Future<Iterable<String>> _getFilteredFilePaths(Stream<File> files) async {
+    // Returns a pattern to check for [directories] as a subset of a file path.
+    RegExp pathFragmentForDirectories(List<String> directories) {
+      final String s = p.separator;
+      return RegExp('(?:^|$s)${p.joinAll(directories)}$s');
+    }
+
+    return files
+        .map((File file) => file.path)
+        .where((String path) =>
+            // Ignore files in build/ directories (e.g., headers of frameworks)
+            // to avoid useless extra work in local repositories.
+            !path.contains(
+                pathFragmentForDirectories(<String>['example', 'build'])) &&
```

Comment:
Can these be implemented with some util from the path library? maybe [isWithin](https://api.flutter.dev/flutter/package-path_path/isWithin.html)? (probably not, it looks like it expects absolute paths)

Reply:
I looked a bit and couldn't find anything that did arbitrary relate path containment.